[FLINK-16946] Update user documentation for job manager memory model (#11947)
apache/flink: docs/ops/memory/mem_setup_jm.md

---
title: "Set up Job Manager Memory"
nav-parent_id: ops_mem
nav-pos: 3
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

* toc
{:toc}

> 

Reviewer: I feel like this is missing an introductory sentence. Something like: 

"Flinks JobManager controls x, y, and z. This guide walks you through high level and fine grained memory configurations for this component.

Reviewer 2: Will add "The Flink Master is the master of the Flink Cluster. 
It consists of three distinct components: Flink Resource Manager, Flink Dispatcher and one Flink JobManager per running Flink Job.
This guide walks you through high level and fine grained memory configurations for the Flink Master."

The further described memory configuration is applicable starting with the release version *1.11*. If you upgrade Flink
from earlier versions, check the [migration guide](mem_migration.html) because many changes were introduced with the *1.11* release.

<span class="label label-info">Note</span> This memory setup guide is relevant <strong>only for job managers</strong>!
The main job manager memory components have a similar but simpler structure compared to the [task executors'](mem_setup_tm.html).

## Configure Total Memory

> If you run the job manager process locally, you do not need to configure memory options, they will have no effect.
> See also [notes for the execution mode](#notes-for-the-execution-mode). Otherwise, the simplest way to set up the
> memory configuration is to configure the total memory.
> See [how to configure it for the Flink process](mem_setup.html#configure-total-memory).

Reviewer: This paragraph felt backwards to me. 
```suggestion
The simplest way to set up the memory configuration is to configure the [total memory](mem_setup.html#configure-total-memory) for the process. 
If you run the job manager process using a local [execution mode](#notes-for-the-execution-mode)., you do not need to configure memory options, they will have no effect.
```

| JVM Overhead                                                   | [`jobmanager.memory.jvm-overhead.min`](../config.html#jobmanager-memory-jvm-overhead-min) <br/> [`jobmanager.memory.jvm-overhead.max`](../config.html#jobmanager-memory-jvm-overhead-max) <br/> [`jobmanager.memory.jvm-overhead.fraction`](../config.html#jobmanager-memory-jvm-overhead-fraction) | Native memory reserved for other JVM overhead: e.g. thread stacks, code cache, garbage collection space etc, it is a [capped fractionated component](mem_setup.html#capped-fractionated-components) of the [total process memory](mem_setup.html#configure-total-memory) |
{:.table-bordered}
<br/>

### Configure JVM Heap

> As mentioned before in the [total memory description](mem_setup.html#configure-total-memory), another way to setup memory
> for job manager is to specify explicitly *JVM Heap* size ([`jobmanager.memory.heap.size`](../config.html#jobmanager-memory-heap-size)).
> It gives more control over the available *JVM Heap* to the following purposes:
> 
> * Flink framework (e.g. *Job cache*)
> * User code running during job submission (e.g. for certain batch sources) or in checkpoint completion callbacks

Reviewer: ? 
```suggestion
As mentioned before in the [total memory description](mem_setup.html#configure-total-memory), another way to setup memory
for job manager is to specify explicitly *JVM Heap* size ([`jobmanager.memory.heap.size`](../config.html#jobmanager-memory-heap-size)).
It gives more control over the available *JVM Heap* which contains:

* Flink framework (e.g. *Job cache*)
* User code running during job submission (e.g. for certain batch sources) or in checkpoint completion callbacks
```

> The required size of *JVM Heap* is mostly driven by the amount of running jobs, their structure and requirements for

Reviewer: ```suggestion
The required size of *JVM Heap* is mostly driven by the number of running jobs, their structure, and requirements for
```

the mentioned user code.

The *Job cache* resides in the *JVM Heap*. It already can be configured currently by
[`jobstore.cache-size`](../config.html#jobstore-cache-size) which must be less than the configured or derived *JVM Heap*.

<span class="label label-info">Note</span> If you have configured the *JVM Heap* explicitly, it is recommended to set
neither *total process memory* nor *total Flink memory*. Otherwise, it may easily lead to memory configuration conflicts.
The *JVM Heap* size is set as the corresponding JVM parameters (*-Xms* and *-Xmx*) when the job manager process is started
by Flink’s scripts or CLI, see also [JVM parameters](mem_setup.html#jvm-parameters).

### Configure Off-heap Memory

The *Off-heap* memory component accounts for any type of *JVM direct memory* and *native memory* usage. Therefore, it
is also set as the corresponding JVM argument: *-XX:MaxDirectMemorySize*, see also [JVM parameters](mem_setup.html#jvm-parameters).

The size of this component can be configured by [`jobmanager.memory.off-heap.size`](../config.html#jobmanager-memory-off-heap-size)

> option. This option can be tuned e.g. if the job manager process throws ‘OutOfMemoryError: Direct buffer memory’, see
> also [the troubleshooting guide](mem_trouble.html#outofmemoryerror-direct-buffer-memory).

Reviewer: ```suggestion
option. This option can be tuned e.g. if the job manager process throws ‘OutOfMemoryError: Direct buffer memory’, see
the [troubleshooting guide](mem_trouble.html#outofmemoryerror-direct-buffer-memory) for more information.
```

There can be the following possible sources of the *Off-heap* memory consumption in JM:

* Flink framework dependencies (e.g. Akka network communication)

> * User code running during job submission (e.g. for certain batch sources) or in checkpoint completion callbacks

Reviewer: ```suggestion
* User code executed during job submission (e.g. for certain batch sources) or in checkpoint completion callbacks
```

> ## Notes for the execution mode

Reviewer 3: I'm wondering whether this section should be titled "local execution" or so, to align with the TM doc.
It is practically talking about local execution as an exception apart from other execution modes.

Reviewer 2: Makes sense to me.

As mentioned before in the [total memory description](mem_setup.html#configure-total-memory), another way to set up the memory
for the Flink Master is to specify explicitly the *JVM Heap* size ([`jobmanager.memory.heap.size`](../config.html#jobmanager-memory-heap-size)).

> It gives more control over the available *JVM Heap* which contains:

Author: ```suggestion
It gives more control over the available *JVM Heap* which is used by:
```
or `which you can use for`
or other lines have to be formulated as some memory components

Reviewer 2: True, I will correct it.
